scvi/external/cellassign/_module.py

```diff
@@ -0,0 +1,188 @@
+
+        # compute phi of NegBin - shape (n_cells, n_genes, n_labels)
+        a = torch.exp(self.log_a)  # (10)
+        a_e = a.expand(s.shape[0], self.n_genes, self.n_labels, self.B)
+        b_init = 2 * ((basis_means_fixed[1] - basis_means_fixed[0]) ** 2)
+        b = torch.exp(torch.ones(self.B) * (-np.log(b_init)))  # (10)
+        b_e = b.expand(s.shape[0], self.n_genes, self.n_labels, self.B)
+        mu_ngc_u = mu_ngc.unsqueeze(-1)
+        mu_ngcb = mu_ngc_u.expand(
+            s.shape[0], self.n_genes, self.n_labels, self.B
+        )  # (128, 100, 5, 10)
+        basis_means_e = basis_means.expand(
+            s.shape[0], self.n_genes, self.n_labels, self.B
+        )  # (128, 100, 5, 10)
+        phi = (  # (128, 100, 5)
+            torch.sum(a_e * torch.exp(-b_e * torch.square(mu_ngcb - basis_means_e)), 3)
+            + self.LOWER_BOUND
+        )
+
+        # compute gamma
+        pdb.set_trace()
+        nb_pdf = NegativeBinomial(probs=mu_ngc, total_count=phi)
+        y_ = x.unsqueeze(-1).expand(s.shape[0], self.n_genes, self.n_labels)
+        y_log_prob_raw = nb_pdf.log_prob(y_)  # (128, 100, 5)
+        theta_log_e = theta_log.expand(128, 5)
+        p_y_unorm = torch.sum(y_log_prob_raw, 1) + theta_log_e  # (128, 5)
+        p_y_norm = torch.logsumexp(p_y_unorm, 1)
+        p_y_norm_e = p_y_norm.unsqueeze(-1).expand(128, 5)
+        gamma = torch.exp(p_y_unorm - p_y_norm_e)  # (128, 5)
+
+        return dict(
+            mu=mu_ngc,
+            phi=phi,
+            gamma=gamma,
+            p_y_norm=p_y_norm,
+        )
+
+    def loss(
+        self,
+        tensors,
+        inference_outputs,
+        generative_outputs,
+        n_obs: int = 1.0,
+    ):
+        # generative_outputs is a dict of the return value from `generative(...)`
+        # assume that `n_obs` is the number of training data points
+        p_y_norm = generative_outputs["p_y_norm"]
+
+        # compute Q
+        # gamma_fixed = torch.empty((None, self.n_labels))
+        # Q = -torch.einsum("nc,cn->", gamma_fixed, p_y_unorm)
+        Q = p_y_norm
+
+        # second term in SCVILoss is Q per cell without prior terms. shape is (n_cells,)
+        loss = torch.sum(p_y_norm)
+
+        # third term is log prob of prior terms in Q
+        log_softmax = torch.nn.LogSoftmax()
+        theta_log = log_softmax(self.theta_logit)
+        theta_log_prior = Dirichlet(torch.tensor(self.dirichlet_concentration))
+        theta_log_prob = -theta_log_prior.log_prob(
+            torch.exp(theta_log) + self.THETA_LOWER_BOUND
+        )
+        delta_log_prior = Normal(self.delta_log_mean, self.delta_log_variance)
+        summed_delta_log = torch.sum(self.delta_log)
+        delta_log_prob = -torch.sum(delta_log_prior.log_prob(summed_delta_log))
+        prior_log_prob = theta_log_prob
+        if self.shrinkage:
+            prior_log_prob += delta_log_prob
+
+        return SCVILoss(loss, Q, prior_log_prob, 0.0)
```

Comment:
```suggestion
        return LossRecorder(loss, Q_per_cell, torch.zeros_like(Q_per_cell), prior_log_prob)
```
Q_per_cell should be the per cell (no outside sum over n) of this value.

And therefore, loss would be `torch.mean(Q_per_cell) * n_obs + prior_log_prob`
![Screen Shot 2021-01-23 at 10 10 30 AM](https://user-images.githubusercontent.com/10859440/105610182-3a86f280-5d63-11eb-9fd3-34eebc14cb14.png)